sql/core/src/main/scala/org/apache/spark/sql/execution/command/FileWritingCommand.scala

```diff
@@ -0,0 +1,93 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.command
+
+import org.apache.spark.SparkContext
+import org.apache.spark.sql.{Row, SparkSession}
+import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow}
+import org.apache.spark.sql.catalyst.plans.logical
+import org.apache.spark.sql.execution.{SparkPlan, SQLExecution}
+import org.apache.spark.sql.execution.datasources.ExecutedWriteSummary
+import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics}
+import org.apache.spark.util.Utils
+
+/**
+ * A logical command specialized for writing data out. `FileWritingCommand`s are
+ * wrapped in `FileWritingCommandExec` during execution.
+ */
+trait FileWritingCommand extends logical.Command {
+  def run(
+    sparkSession: SparkSession,
+    children: Seq[SparkPlan],
+    fileCommandExec: FileWritingCommandExec): Seq[Row]
+}
+
+/**
+ * A physical operator specialized to execute the run method of a `FileWritingCommand`,
+ * save the result to prevent multiple executions, and record necessary metrics for UI.
+ */
+case class FileWritingCommandExec(
+    cmd: FileWritingCommand,
+    children: Seq[SparkPlan],
+    givenMetrics: Option[Map[String, SQLMetric]] = None) extends CommandExec {
+
+  override val metrics = givenMetrics.getOrElse {
+    val sparkContext = sqlContext.sparkContext
+    Map(
+      // General metrics.
+      "avgTime" -> SQLMetrics.createMetric(sparkContext, "average writing time (ms)"),
+      "numFiles" -> SQLMetrics.createMetric(sparkContext, "number of written files"),
+      "numOutputBytes" -> SQLMetrics.createMetric(sparkContext, "bytes of written output"),
+      "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
+      "numParts" -> SQLMetrics.createMetric(sparkContext, "number of dynamic part")
+    )
+  }
+
+  /**
+   * Callback function that update metrics collected from the writing operation.
+   */
+  private[sql] def postDriverMetrics(writeSummaries: Seq[ExecutedWriteSummary]): Unit = {
+    var numPartitions = 0
+    var numFiles = 0
+    var totalNumBytes: Long = 0L
+    var totalNumOutput: Long = 0L
+
+    writeSummaries.foreach { summary =>
+      numPartitions += summary.updatedPartitions.size
+      numFiles += summary.numOutputFile
+      totalNumBytes += summary.numOutputBytes
+      totalNumOutput += summary.numOutputRows
+    }
+
+    // The time for writing individual file can be zero if it's less than 1 ms. Zero values can
+    // lower actual time of writing to zero when calculating average, so excluding them.
+    val avgWritingTime =
+      Utils.average(writeSummaries.flatMap(_.writingTimePerFile.filter(_ > 0))).toLong
+    // Note: for simplifying metric values assignment, we put the values as the alphabetically
+    // sorted of the metric keys.
+    val metricsNames = metrics.keys.toSeq.sorted
```

Comment:
how do you guarantee the `metrics` contains avg, numFiles, etc. as it's created by `givenMetrics.getOrElse`?